RFC-0004-support-distributed-procedure.md

```diff
@@ -0,0 +1,397 @@
+public boolean isDistributedProcedure(ConnectorId connectorId, SchemaTableName name);
+```
+
+Extract interface `IProcedureRegistry` for `ProcedureRegistry` and declare it in `presto-spi` module, so that it can be recognized and used in both the `presto-analyzer` and `connector` modules.
+
+```java
+public interface IProcedureRegistry
+{
+    void addProcedures(ConnectorId connectorId, Collection<Procedure> procedures);
+
+    void removeProcedures(ConnectorId connectorId);
+
+    Procedure resolve(ConnectorId connectorId, SchemaTableName name);
+
+    DistributedProcedure resolveDistributed(ConnectorId connectorId, SchemaTableName name);
+
+    boolean isDistributedProcedure(ConnectorId connectorId, SchemaTableName name);
+}
+```
+
+#### 2. Set new query type during preparer phase
+
+Add a new type in `QueryType` called `CALL_DISTRIBUTED_PROCEDURE(10)` for `call distributed procedure` statement, it's different from the query type 'DATA-DEFINITION' originally set for `call procedure` statement.
+
+In the preparer phase, when encountering a call statement, utilize the injected `IProcedureRegistry` to determine whether the called procedure is a `DistributedProcedure`. If so, associate its query type to `CALL_DISTRIBUTED_PROCEDURE`.
+
+In this way, in `LocalDispatchQueryFactory` we can choose `SqlQueryExecutionFactory` to generate the query execution for `call distributed procedure` statement. This allows for the utilization of distributed data scanning, writing, and unified committing mechanism during actual execution.
+
+#### 3. Processing logic during analyzing and logical planning phase
+
+Add a new plan node type `CallDistributedProcedureNode` for `call distributed procedure` statement, which is defined as follows:
+
+```java
+public class CallDistributedProcedureNode
+            extends InternalPlanNode
+{
+    private final PlanNode source;
+    private final Optional<CallDistributedProcedureTarget> target;
+    private final VariableReferenceExpression rowCountVariable;
+    private final VariableReferenceExpression fragmentVariable;
+    private final VariableReferenceExpression tableCommitContextVariable;
+    private final List<VariableReferenceExpression> columns;
+    private final List<String> columnNames;
+    private final Set<VariableReferenceExpression> notNullColumnVariables;
+    private final Optional<PartitioningScheme> partitioningScheme;
+    private final List<VariableReferenceExpression> outputs;
+
+    ......
+}
+```
+
+At present, it can be regarded as a simplified version of `TableWriterNode`, and holds `CallDistributedProcedureTarget`, a subclass of `WriterTarget` used to maintain information about the procedure and the call parameters:
+
+```java
+public static class CallDistributedProcedureTarget
+            extends WriterTarget
+{
+    private final QualifiedObjectName procedureName;
+    private final Object[] procedureArguments;
```

Comment:
Presume the procedureArguments have types. There could be a difference in the types supported by native vs java execution. Would be great to clarify about that.

Reply:
As I understand, `CallDistributedProcedureTarget` is a subclass of `WriterTarget` which is used only in coordinator. That is, it's always used in java execution environment.

`ExecuteProcedureHandle` which is a subclass of `ExecutionWriterTarget` will be sent to worker. The handling of it is similar to `InsertHandle` or `CreateHandle`.